File: langs/python/src/dotlyte/validator.py
"""Schema validation system for DOTLYTE.

Validates configuration against a user-defined schema with type checking,
required fields, format validation, enums, min/max, and custom validators.
"""

from __future__ import annotations

import re
from typing import Any, Callable, Literal

from dotlyte.errors import ValidationError

# ──── Schema Types ────

SchemaType = Literal["string", "number", "boolean", "array", "object"]

SchemaFormat = Literal[
    "url", "email", "ip", "ipv4", "ipv6",
    "hostname", "port", "uuid", "date", "iso-date",
]


class SchemaRule:
    """A single field's validation rule."""

    def __init__(
        self,
        *,
        type: SchemaType | None = None,
        required: bool = False,
        format: SchemaFormat | str | None = None,
        enum: list[Any] | None = None,
        min: float | int | None = None,
        max: float | int | None = None,
        default: Any = None,
        sensitive: bool = False,
        doc: str | None = None,
        validator: Callable[[Any], bool] | None = None,
    ) -> None:
        self.type = type
        self.required = required
        self.format = format
        self.enum = enum
        self.min = min
        self.max = max
        self.default = default
        self.sensitive = sensitive
        self.doc = doc
        self.validator = validator


DotlyteSchema = dict[str, SchemaRule]


class SchemaViolation:
    """A single validation failure."""

    def __init__(
        self,
        key: str,
        rule: str,
        message: str,
        expected: Any = None,
        actual: Any = None,
    ) -> None:
        self.key = key
        self.rule = rule
        self.message = message
        self.expected = expected
        self.actual = actual

    def __repr__(self) -> str:
        return f"SchemaViolation(key={self.key!r}, rule={self.rule!r}, message={self.message!r})"


# ──── Built-in Format Validators ────

_FORMAT_VALIDATORS: dict[str, Callable[[str], bool]] = {
    "url": lambda v: bool(re.match(r"^https?://\S+", v)),
    "email": lambda v: bool(re.match(r"^[^@\s]+@[^@\s]+\.[^@\s]+$", v)),
    "ip": lambda v: _is_ipv4(v) or _is_ipv6(v),
    "ipv4": _is_ipv4 if False else lambda v: _is_ipv4(v),
    "ipv6": lambda v: _is_ipv6(v),
    "hostname": lambda v: bool(re.match(r"^[a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?(\.[a-zA-Z]{2,})+$", v)) or v == "localhost",
    "port": lambda v: v.isdigit() and 1 <= int(v) <= 65535,
    "uuid": lambda v: bool(re.match(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", v, re.I)),
    "date": lambda v: bool(re.match(r"^\d{4}-\d{2}-\d{2}$", v)),
    "iso-date": lambda v: bool(re.match(r"^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:?\d{2})?)?$", v)),
}


def _is_ipv4(v: str) -> bool:
    parts = v.split(".")
    if len(parts) != 4:
        return False
    try:
        return all(0 <= int(p) <= 255 for p in parts)
    except ValueError:
        return False


def _is_ipv6(v: str) -> bool:
    try:
        import ipaddress
        ipaddress.IPv6Address(v)
        return True
    except (ValueError, AttributeError):
        return False


# ──── Validation Functions ────

def apply_schema_defaults(
    data: dict[str, Any],
    schema: DotlyteSchema,
) -> dict[str, Any]:
    """Apply schema default values to missing keys.

    Args:
        data: The current configuration data.
        schema: The schema definition.

    Returns:
        A new dict with defaults filled in.

    """
    result = data.copy()
    for key, rule in schema.items():
        if key not in result and rule.default is not None:
            result[key] = rule.default
    return result


def validate_schema(
    data: dict[str, Any],
    schema: DotlyteSchema,
    strict: bool = False,
) -> list[SchemaViolation]:
    """Validate data against a schema. Returns list of violations.

    Args:
        data: Configuration data to validate.
        schema: Schema to validate against.
        strict: If True, unknown keys are rejected.

    Returns:
        List of SchemaViolation objects (empty = valid).

    """
    violations: list[SchemaViolation] = []

    for key, rule in schema.items():
        value = data.get(key)

        # Required check
        if rule.required and (value is None or key not in data):
            violations.append(SchemaViolation(
                key=key,
                rule="required",
                message=f"Required key '{key}' is missing",
            ))
            continue

        if value is None or key not in data:
            continue

        # Type check
        if rule.type and not _check_type(value, rule.type):
            violations.append(SchemaViolation(
                key=key,
                rule="type",
                message=f"Key '{key}' must be of type {rule.type}, got {type(value).__name__}",
                expected=rule.type,
                actual=type(value).__name__,
            ))

        # Enum check
        if rule.enum is not None and value not in rule.enum:
            violations.append(SchemaViolation(
                key=key,
                rule="enum",
                message=f"Key '{key}' must be one of {rule.enum}, got {value!r}",
                expected=rule.enum,
                actual=value,
            ))

        # Min/max checks
        if rule.min is not None and isinstance(value, (int, float)):
            if value < rule.min:
                violations.append(SchemaViolation(
                    key=key,
                    rule="min",
                    message=f"Key '{key}' must be >= {rule.min}, got {value}",
                    expected=rule.min,
                    actual=value,
                ))

        if rule.max is not None and isinstance(value, (int, float)):
            if value > rule.max:
                violations.append(SchemaViolation(
                    key=key,
                    rule="max",
                    message=f"Key '{key}' must be <= {rule.max}, got {value}",
                    expected=rule.max,
                    actual=value,
                ))

        # Format check
        if rule.format and isinstance(value, str):
            fmt = rule.format
            if fmt in _FORMAT_VALIDATORS:
                if not _FORMAT_VALIDATORS[fmt](value):
                    violations.append(SchemaViolation(
                        key=key,
                        rule="format",
                        message=f"Key '{key}' doesn't match format '{fmt}'",
                        expected=fmt,
                        actual=value,
                    ))
            else:
                # Treat as regex
                if not re.search(fmt, value):
                    violations.append(SchemaViolation(
                        key=key,
                        rule="format",
                        message=f"Key '{key}' doesn't match pattern '{fmt}'",
                        expected=fmt,
                        actual=value,
                    ))

        # Custom validator
        if rule.validator is not None:
            try:
                if not rule.validator(value):
                    violations.append(SchemaViolation(
                        key=key,
                        rule="validator",
                        message=f"Key '{key}' failed custom validation",
                        actual=value,
                    ))
            except Exception as e:
                violations.append(SchemaViolation(
                    key=key,
                    rule="validator",
                    message=f"Key '{key}' custom validator threw: {e}",
                    actual=value,
                ))

    # Strict mode: reject unknown keys
    if strict:
        schema_keys = set(schema.keys())
        for key in data:
            if key not in schema_keys:
                violations.append(SchemaViolation(
                    key=key,
                    rule="strict",
                    message=f"Unknown key '{key}' not in schema (strict mode)",
                ))

    return violations


def assert_valid(
    data: dict[str, Any],
    schema: DotlyteSchema,
    strict: bool = False,
) -> None:
    """Validate data against schema, raising ValidationError if invalid.

    Args:
        data: Configuration data to validate.
        schema: Schema to validate against.
        strict: If True, unknown keys are rejected.

    Raises:
        ValidationError: If validation fails.

    """
    violations = validate_schema(data, schema, strict)
    if violations:
        raise ValidationError(violations)


def get_sensitive_keys(schema: DotlyteSchema) -> set[str]:
    """Extract keys marked as sensitive in the schema.

    Args:
        schema: Schema definition.

    Returns:
        Set of key names marked as sensitive.

    """
    return {key for key, rule in schema.items() if rule.sensitive}


def _check_type(value: Any, expected: SchemaType) -> bool:
    """Check if a value matches the expected schema type."""
    if expected == "string":
        return isinstance(value, str)
    if expected == "number":
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    if expected == "boolean":
        return isinstance(value, bool)
    if expected == "array":
        return isinstance(value, list)
    if expected == "object":
        return isinstance(value, dict)
    return True

File: langs/javascript/src/validator.ts
/**
 * Schema validation engine for DOTLYTE.
 *
 * Validates config values against a user-defined schema. Collects ALL
 * violations before throwing — never fail-on-first.
 *
 * Supports: required, type, format, enum, min/max, sensitive, doc, default.
 *
 * @example
 * ```ts
 * const schema: DotlyteSchema = {
 *   port:     { type: "number", required: true, min: 1, max: 65535, doc: "Server port" },
 *   host:     { type: "string", format: "hostname" },
 *   debug:    { type: "boolean", default: false },
 *   api_key:  { type: "string", required: true, sensitive: true },
 *   log:      { type: "string", enum: ["debug","info","warn","error"] },
 * };
 * ```
 */

import { ValidationError, type SchemaViolation } from "./errors.js";

/* ──────── Schema definition types ──────── */

export type SchemaType = "string" | "number" | "boolean" | "array" | "object";

export type SchemaFormat =
  | "url"
  | "email"
  | "ip"
  | "ipv4"
  | "ipv6"
  | "hostname"
  | "port"
  | "uuid"
  | "date"
  | "iso-date"
  | (string & {}); // allows arbitrary regex patterns

export interface SchemaRule {
  /** Expected type after coercion. */
  type?: SchemaType;
  /** Key is mandatory — must be present and non-null. */
  required?: boolean;
  /** Built-in format name or regex pattern string. */
  format?: SchemaFormat;
  /** List of allowed values. */
  enum?: unknown[];
  /** Minimum value (numbers) or length (strings/arrays). */
  min?: number;
  /** Maximum value (numbers) or length (strings/arrays). */
  max?: number;
  /** Default value if key is missing. Applied before validation. */
  default?: unknown;
  /** Mark this key as sensitive — will be redacted in toString/toJSON. */
  sensitive?: boolean;
  /** Human-readable description. Self-documenting config. */
  doc?: string;
  /** Custom validator function. Return string error or null for pass. */
  validator?: (value: unknown) => string | null;
}

/** Full schema: a map of dot-notation keys to rules. */
export type DotlyteSchema = Record<string, SchemaRule>;

/* ──────── Built-in format validators ──────── */

const FORMAT_VALIDATORS: Record<string, (v: string) => boolean> = {
  url: (v) => {
    try {
      new URL(v);
      return true;
    } catch {
      return false;
    }
  },
  email: (v) => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(v),
  ip: (v) => FORMAT_VALIDATORS.ipv4!(v) || FORMAT_VALIDATORS.ipv6!(v),
  ipv4: (v) =>
    /^(\d{1,3}\.){3}\d{1,3}$/.test(v) &&
    v.split(".").every((n) => parseInt(n) >= 0 && parseInt(n) <= 255),
  ipv6: (v) => /^([0-9a-fA-F]{0,4}:){2,7}[0-9a-fA-F]{0,4}$/.test(v),
  hostname: (v) => /^[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?(\.[a-zA-Z]{2,})*$/.test(v),
  port: (v) => {
    const n = parseInt(v, 10);
    return !isNaN(n) && n >= 1 && n <= 65535;
  },
  uuid: (v) =>
    /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(v),
  date: (v) => !isNaN(Date.parse(v)),
  "iso-date": (v) => /^\d{4}-\d{2}-\d{2}(T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:?\d{2})?)?$/.test(v),
};

/* ──────── Core validation ──────── */

/**
 * Apply default values from schema to the config data.
 * Returns a new object — does not mutate input.
 */
export function applySchemaDefaults(
  data: Record<string, unknown>,
  schema: DotlyteSchema,
): Record<string, unknown> {
  const result = { ...data };

  for (const [key, rule] of Object.entries(schema)) {
    if (rule.default !== undefined) {
      const current = getNestedValue(result, key);
      if (current === undefined || current === null) {
        setNestedValue(result, key, rule.default);
      }
    }
  }

  return result;
}

/**
 * Validate config data against a schema.
 *
 * @param data   — the merged, coerced config object
 * @param schema — the schema rules
 * @param strict — if true, reject keys NOT in the schema (catch typos)
 * @returns Array of violations (empty = valid)
 */
export function validateSchema(
  data: Record<string, unknown>,
  schema: DotlyteSchema,
  strict = false,
): SchemaViolation[] {
  const violations: SchemaViolation[] = [];

  // Validate declared keys
  for (const [key, rule] of Object.entries(schema)) {
    const value = getNestedValue(data, key);

    // Required check
    if (rule.required && (value === undefined || value === null)) {
      violations.push({
        key,
        rule: "required",
        message: `Key '${key}' is required but missing or null.${rule.doc ? ` (${rule.doc})` : ""}`,
      });
      continue; // Skip further checks on missing required key
    }

    // Skip optional keys that are absent
    if (value === undefined || value === null) continue;

    // Type check
    if (rule.type) {
      if (!checkType(value, rule.type)) {
        violations.push({
          key,
          rule: "type",
          message: `Key '${key}' expected type '${rule.type}', got '${typeof value}'.`,
          expected: rule.type,
          actual: typeof value,
        });
      }
    }

    // Format check (only for strings)
    if (rule.format && typeof value === "string") {
      const builtin = FORMAT_VALIDATORS[rule.format];
      if (builtin) {
        if (!builtin(value)) {
          violations.push({
            key,
            rule: "format",
            message: `Key '${key}' does not match format '${rule.format}'. Got: '${value}'.`,
            expected: rule.format,
            actual: value,
          });
        }
      } else {
        // Treat as regex pattern
        try {
          if (!new RegExp(rule.format).test(value)) {
            violations.push({
              key,
              rule: "format",
              message: `Key '${key}' does not match pattern '${rule.format}'.`,
              expected: rule.format,
              actual: value,
            });
          }
        } catch {
          violations.push({
            key,
            rule: "format",
            message: `Invalid format pattern '${rule.format}' for key '${key}'.`,
          });
        }
      }
    }

    // Enum check
    if (rule.enum && !rule.enum.includes(value)) {
      violations.push({
        key,
        rule: "enum",
        message: `Key '${key}' must be one of [${rule.enum.join(", ")}], got '${value}'.`,
        expected: rule.enum,
        actual: value,
      });
    }

    // Min/max (numbers)
    if (typeof value === "number") {
      if (rule.min !== undefined && value < rule.min) {
        violations.push({
          key,
          rule: "min",
          message: `Key '${key}' value ${value} is below minimum ${rule.min}.`,
          expected: rule.min,
          actual: value,
        });
      }
      if (rule.max !== undefined && value > rule.max) {
        violations.push({
          key,
          rule: "max",
          message: `Key '${key}' value ${value} exceeds maximum ${rule.max}.`,
          expected: rule.max,
          actual: value,
        });
      }
    }

    // Min/max (strings — length)
    if (typeof value === "string") {
      if (rule.min !== undefined && value.length < rule.min) {
        violations.push({
          key,
          rule: "min",
          message: `Key '${key}' length ${value.length} is below minimum ${rule.min}.`,
        });
      }
      if (rule.max !== undefined && value.length > rule.max) {
        violations.push({
          key,
          rule: "max",
          message: `Key '${key}' length ${value.length} exceeds maximum ${rule.max}.`,
        });
      }
    }

    // Min/max (arrays — length)
    if (Array.isArray(value)) {
      if (rule.min !== undefined && value.length < rule.min) {
        violations.push({
          key,
          rule: "min",
          message: `Key '${key}' has ${value.length} items, minimum is ${rule.min}.`,
        });
      }
      if (rule.max !== undefined && value.length > rule.max) {
        violations.push({
          key,
          rule: "max",
          message: `Key '${key}' has ${value.length} items, maximum is ${rule.max}.`,
        });
      }
    }

    // Custom validator
    if (rule.validator) {
      const err = rule.validator(value);
      if (err) {
        violations.push({ key, rule: "validator", message: err });
      }
    }
  }

  // Strict mode: reject keys not in schema
  if (strict) {
    const schemaKeys = new Set(Object.keys(schema));
    const allDataKeys = flattenKeys(data);
    for (const key of allDataKeys) {
      if (!schemaKeys.has(key)) {
        violations.push({
          key,
          rule: "strict",
          message: `Unknown config key '${key}' — not defined in schema. Possible typo?`,
        });
      }
    }
  }

  return violations;
}

/**
 * Validate and throw if violations exist.
 */
export function assertValid(
  data: Record<string, unknown>,
  schema: DotlyteSchema,
  strict = false,
): void {
  const violations = validateSchema(data, schema, strict);
  if (violations.length > 0) {
    throw new ValidationError(violations);
  }
}

/**
 * Extract the set of sensitive keys from a schema.
 */
export function getSensitiveKeys(schema: DotlyteSchema): Set<string> {
  const keys = new Set<string>();
  for (const [key, rule] of Object.entries(schema)) {
    if (rule.sensitive) keys.add(key);
  }
  return keys;
}

/* ──────── Helpers ──────── */

function checkType(value: unknown, expected: SchemaType): boolean {
  switch (expected) {
    case "string":
      return typeof value === "string";
    case "number":
      return typeof value === "number" && !isNaN(value);
    case "boolean":
      return typeof value === "boolean";
    case "array":
      return Array.isArray(value);
    case "object":
      return typeof value === "object" && value !== null && !Array.isArray(value);
    default:
      return true;
  }
}

function getNestedValue(obj: Record<string, unknown>, key: string): unknown {
  const parts = key.split(".");
  let current: unknown = obj;
  for (const part of parts) {
    if (current !== null && typeof current === "object" && !Array.isArray(current)) {
      current = (current as Record<string, unknown>)[part];
    } else {
      return undefined;
    }
  }
  return current;
}

function setNestedValue(obj: Record<string, unknown>, key: string, value: unknown): void {
  const parts = key.split(".");
  let current = obj;
  for (let i = 0; i < parts.length - 1; i++) {
    const part = parts[i]!;
    if (!(part in current) || typeof current[part] !== "object") {
      current[part] = {};
    }
    current = current[part] as Record<string, unknown>;
  }
  current[parts[parts.length - 1]!] = value;
}

function flattenKeys(obj: Record<string, unknown>, prefix = ""): string[] {
  const keys: string[] = [];
  for (const [key, value] of Object.entries(obj)) {
    const full = prefix ? `${prefix}.${key}` : key;
    if (value !== null && typeof value === "object" && !Array.isArray(value)) {
      keys.push(...flattenKeys(value as Record<string, unknown>, full));
    } else {
      keys.push(full);
    }
  }
  return keys;
}
